walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> allAttendees = request.getAllAttendees();
+        long duration = request.getDuration();
+
+        Collection<Event> mandatoryAttendeeEvents = getEventsForAttendees(mandatoryAttendees, events);
+        ArrayList<TimeRange> mandatoryAttendeesSlots = getPossibleSlots(mandatoryAttendeeEvents, mandatoryAttendees, duration);
+
+        Collection<Event> allAttendeeEvents = getEventsForAttendees(allAttendees, events);
+        ArrayList<TimeRange> allAttendeeSlots = getPossibleSlots(allAttendeeEvents, allAttendees, duration);
+
+        if (allAttendeeSlots.size() == 0 && mandatoryAttendees.size() != 0) {
+            return mandatoryAttendeesSlots;
+        }
+        else {
+            return allAttendeeSlots;
+        }
+
+    }
+
+    public Collection<Event> getEventsForAttendees(Collection<String> attendees, Collection<Event> events) {
+        ArrayList<Event> filteredEvents = new ArrayList<>();;
+        for (Event event : events) {
+            Set<String> eventAttendees = event.getAttendees();
+            for (String attendee : attendees) {
+                if (eventAttendees.contains(attendee)) {
+                    filteredEvents.add(event);
+                    break;
+                }
+            }
+        }
+        return filteredEvents;
+    }
+
+    public ArrayList<TimeRange> getPossibleSlots(Collection<Event> events, Collection<String> attendees, long duration) {
+        ArrayList<TimeRange> slots = new ArrayList<TimeRange>();
+        int attendeeNumber = attendees.size();
+
+        if (attendeeNumber == 0) {
+            slots.add(TimeRange.WHOLE_DAY);
+            return slots;
+        }
+        else if (duration > TimeRange.WHOLE_DAY.duration()) {
+            return slots;
+        }
```

Comment:
This seems like it should be checked in query. In general, precondition checks that can help short circuit logic (for example duration > WHOLE_DAY returns empty, no attendees returns WHOLE_DAY, etc) are better done as soon as possible so that downstream logic doesn't have to worry about those edge cases